python/rainbow/simulators/prox_soft_bodies/jacobi_parallel.py

```diff
@@ -0,0 +1,143 @@
+        x_b = x[block]
+        b_b = b[block]
+        z_b = z[block]
+        
+        # Solve:         x_n = prox_{R^+}( x_n - r (A x_n + b) )
+        x_b[0] = np.max(np.array([0.0, z_b[0]]))
+
+        # Solve:         x_f = prox_C( x_f - r (A x_f + b))
+        x_b_0 = x_b[0] if hybrid else x_old[block][0]
+        x_b[1], x_b[2], x_b[3] = friction_solver(z_b[1], z_b[2], z_b[3], mu_k, x_b_0)
+        # Put updated contact forces back into solution vector
+        x[block] = x_b
+    return x
+
+
+def solve(J, WJT, b, mu, friction_solver, engine, stats, debug_on, prefix):
+    """ Jacobi Proximal Solver, which just supports parallel computing.
+
+    Args:
+        J (ArrayLike): The contact jacobi matrix.
+        WJT (ArrayLike): The WJ^T matrix, here W = M^{-1}.
+        b (ArrayLike): b = Ju^t + ∆t JM^{-1}h + EJu^t
+        mu (float): The coefficient of friction.
+        friction_solver (callable): The friction cone function.
+        engine (Object): The engine object.
+        stats (dict): The statistics information.
+        debug_on (boolean): Whether to debug.
+        prefix (string): The prefix of the statistics information.
+
+    Returns:
+        (ArrayLike, dict): The new contact force and the statistics information.
+    """
+    timer = None
+    if debug_on:
+        timer = Timer("Gauss Seidel")
+        stats[prefix + "residuals"] = (
+            np.ones(engine.params.max_iterations, dtype=np.float64) * np.inf
+        )
+        stats[prefix + "lambda"] = np.zeros(
+            [engine.params.max_iterations] + list(b.shape), dtype=np.float64
+        )
+        stats[prefix + "reject"] = np.zeros(engine.params.max_iterations, dtype=bool)
+        stats[prefix + "exitcode"] = 0
+        stats[prefix + "iterations"] = engine.params.max_iterations
+        timer.start()
+
+    K = len(engine.contact_points)
+    x = np.zeros(b.shape, dtype=np.float64)  # The current iterate
+    sol = np.zeros(
+        b.shape, dtype=np.float64
+    )  # The last best known solution, used for restarting if divergence
+    error = np.zeros(b.shape, dtype=np.float64)  # The residual vector
+
+    # Compute initial r-factor value
+    delassus_diag = np.sum(J.multiply(WJT.T), axis=1).A1
+    delassus_diag[delassus_diag == 0] = 1
+    r = 0.1 / delassus_diag
+
+    # Extract parameter values for controlling the adaptive r-factor strategy
+    nu_reduce = engine.params.nu_reduce
+    nu_increase = engine.params.nu_increase
+    too_small_merit_change = engine.params.too_small_merit_change
+    last_merit = np.Inf
+
+    # Set number of threads for Numba
+    nb.set_num_threads(4 if (nb.config.NUMBA_NUM_THREADS // 2) >= 4 else nb.config.NUMBA_NUM_THREADS // 2)
+    
+    for iteration in range(engine.params.max_iterations):
+        w = WJT.dot(x)
+        z = x - np.multiply(r, (J.dot(w) + b))
+        hybrid = True if engine.params.proximal_solver['scheme'] == 'jacboi_hybrid' else False
```

Comment:
It is not quite clear what the hybrid setting does

Why set this value inside the for loop? It does not depend on the iterations, it would be more efficient to set it outside the for loop.